Adds command-line parsing package to determine service running mode.

nielsabels opened the pull request:
Adds a NuGet package and an abstraction to determine how the service should launch itself.

NuGet package: https://github.com/gsscoder/commandline
Abstraction: `ApplicationLaunch` class providing en enum of type `StartupMode` that can have the value `RunNormally` or `RunPreflightCheck`.

Commit:
Adds NuGet package.

Commit:
Adds ApplicationLaunch class to determine running mode.

sandermvanvliet commented on ApplicationStartup.cs:
```diff
@@ -0,0 +1,37 @@
+
+    }
+
+    public enum StartupMode
```
Maybe an idea to make this `Normal` and `PreflightCheck` as both modes will run the application?

sandermvanvliet commented on ApplicationStartup.cs:
```diff
@@ -0,0 +1,37 @@
+            }
+            catch(Exception e)
+            {
+                Console.WriteLine("Encountered an exception trying to determine StartupMode [RunNormally, RunPreflightCheck].");
```
Consider using `Console.Error.WriteLine` as on non-Windowses this will output to `STDERR` and that forces it through the normal output redirection that might happen inside Docker/Octo/TeamCity.

sandermvanvliet commented on ApplicationStartup.cs:
```diff
@@ -0,0 +1,37 @@
+            catch(Exception e)
+            {
+                Console.WriteLine("Encountered an exception trying to determine StartupMode [RunNormally, RunPreflightCheck].");
+                Console.WriteLine(e.ToString());
```
Maybe also add `Console.Error.WriteLine(options.GetUsage());` to show the options, that way you don't have to specify that yourself (it gets them from the Options class.)

sandermvanvliet commented on CommandLineOptions.cs:
```diff
@@ -0,0 +1,15 @@
+    public class CommandLineOptions
+    {
+        [Option(HelpText = "Starts the application in a pre-flight mode to validate whether application is good to go.")]
+        public bool Preflight { get; set; }
```
Nitpicking here: as you use `PreflightCheck` everywhere else maybe you can make the name of this property the same? (You can still specify the argument to be `--preflight` through the `OptionAttribute`)

sandermvanvliet commented on MyService.cs:
```diff
@@ -13,34 +13,32 @@ public void Stop()
             else
             {
-                Console.WriteLine("Pre-flight mode wasn't specified. Running normally.");
+                Console.WriteLine("Pre-flight check failed.");
```
Possibly use `Console.Error.WriteLine` here as well (see comment above).

sandermvanvliet commented on MyService.cs:
```diff
@@ -13,34 +13,32 @@ public void Stop()
             }
+
+            Console.WriteLine("Type the following command on your Powershell terminal to find the exit code of the most recently exited process:");
+            Console.WriteLine("echo $LastExitCode");
```
On Windows this is: `echo %errorlevel%`
On OS/X and Linux this is: `echo $?`

sandermvanvliet reviewed (commented):
Looks a lot cleaner now!

nabels-coolblue commented:
Great tips! Cheers for the review. Will address the comments.

Commit:
Addresses feedback.

Renames StartupMode enum members to be more descriptive and clear.

using Console.Error.WriteLine as on non-Windowses this will output to STDERR and that forces it through the normal output redirection that might happen inside Docker/Octo/TeamCity.

Commit:
Removed misleading message.

In case somebody specifies --preflight=asdf, the logging message wrongfully states the service will run in pre-flight mode. That decision is not made by whitelisting the argument, but rather it's made by the ApplicationStartup class.

Commit:
Use Console.Error.Writeline to log errors.

Commit:
Giving different methods to check exit codes.

Commit:
Adds support for checking exit code in Powershell terminal.

Commit:
Updates application output.

nabels-coolblue commented:
@sandermvanvliet: addressed your comments. Please have a look. 

Also updated the README with some more sample output.

Commit:
Updates README with exit codes how-tos.
